source/common/secret/secret_manager_impl.cc

```diff
@@ -2,27 +2,46 @@
 
 #include "envoy/common/exception.h"
 
+#include "common/protobuf/utility.h"
 #include "common/ssl/tls_certificate_config_impl.h"
 
 namespace Envoy {
 namespace Secret {
 
-void SecretManagerImpl::addOrUpdateSecret(const envoy::api::v2::auth::Secret& secret) {
+void SecretManagerImpl::addStaticSecret(const envoy::api::v2::auth::Secret& secret) {
   switch (secret.type_case()) {
-  case envoy::api::v2::auth::Secret::TypeCase::kTlsCertificate:
-    tls_certificate_secrets_[secret.name()] =
+  case envoy::api::v2::auth::Secret::TypeCase::kTlsCertificate: {
+    static_tls_certificate_secrets_[secret.name()] =
         std::make_unique<Ssl::TlsCertificateConfigImpl>(secret.tls_certificate());
     break;
+  }
   default:
     throw EnvoyException("Secret type not implemented");
   }
 }
 
 const Ssl::TlsCertificateConfig*
-SecretManagerImpl::findTlsCertificate(const std::string& name) const {
-  auto secret = tls_certificate_secrets_.find(name);
-  return (secret != tls_certificate_secrets_.end()) ? secret->second.get() : nullptr;
+SecretManagerImpl::findStaticTlsCertificate(const std::string& name) const {
+  auto secret = static_tls_certificate_secrets_.find(name);
+  return (secret != static_tls_certificate_secrets_.end()) ? secret->second.get() : nullptr;
+}
+
+DynamicTlsCertificateSecretProviderSharedPtr
+SecretManagerImpl::findOrCreateDynamicTlsCertificateSecretProvider(
```

Comment:
Still don't understand how these are going to get cleaned up. When listeners/clusters that refer to secrets are removed we are going to leak weak pointers. Do you plan on handling this somehow? Can we handle this in this change? There are various methods to handle but I'm not sure what the long range plan is here.

Reply:
Yes, we may leak weak pointers.  But weak pointer is not using too much space, beside number of sds_config_sources will not be too big, it should be less than 100, maybe it is OK to leak these map entries?   
Well, we will think about how to delete them if it is easy.

Reply:
Sorry leaking them is not an option. One way to deal with this is to occasionally have the manager go through and try to lock all the pointers and if they are null discard them. Please fix.

Reply:
OK, will do.  

Based on the assumption of the map element number is small, in each findOrCreate() call, we will go through all map elements, and remove the ones with o ref_count.

Reply:
Yup I think that's fine.